tests/SecurityGroup.py

```diff
@@ -0,0 +1,127 @@
+        for node in nodes:
+            if node not in self.nonParticipants:
+                for nonParticipant in self.nonParticipants:
+                    match = "" if node == nonParticipant else " not"
+                    Utils.Print("node: '{}' does{} match '{}'".format(node, match, nonParticipant))
+            assert node in self.nonParticipants, "Cannot remove {} from nonParticipants list: {}".format(node, SecurityGroup.createAction(self.nonParticipants))
+            self.nonParticipants.remove(node)
+
+        self.participants.extend(nodes)
+        if Utils.Debug: Utils.Print("nonParticipants list: {}, to the participants list: {}".format(SecurityGroup.createAction(self.nonParticipants), SecurityGroup.createAction(self.participants)))
+        self.__consistentStorage()
+
+    def __remParticipants(self, nodes):
+        if Utils.Debug: Utils.Print("Moving the following: {}, from the participants list: {}, to the non-participants list: {}".format(SecurityGroup.createAction(nodes), SecurityGroup.createAction(self.participants), SecurityGroup.createAction(self.nonParticipants)))
+        for node in nodes:
+            self.participants.remove(node)
+
+        self.nonParticipants.extend(nodes)
+        if Utils.Debug: Utils.Print("participants list: {}, to the non-participants list: {}".format(SecurityGroup.createAction(self.participants), SecurityGroup.createAction(self.nonParticipants)))
+        self.__consistentStorage()
+
+    @staticmethod
+    def createAction(nodes):
+        return None if len(nodes) == 0 else \
+            "[[{}]]".format(','.join(['"{}"'.format(node.getParticipant()) for node in nodes]))
+
+    def editSecurityGroup(self, addNodes=[], removeNodes=[]):
+
+        addAction = SecurityGroup.createAction(addNodes)
+        # doing deep copy in case the passed in list IS participants or nonParticipants lists, which will be adjusted
+        self.__addParticipants(copy.deepcopy(addNodes))
+        removeAction = SecurityGroup.createAction(removeNodes)
+        self.__remParticipants(copy.deepcopy(removeNodes))
+
+        if addAction:
+            Utils.Print("adding {} to the security group".format(addAction))
+            trans = self.defaultNode.pushMessage(self.contractAccount.name, "add", addAction, "--permission eosio@active")
+            Utils.Print("add trans: {}".format(json.dumps(trans, indent=4, sort_keys=True)))
+
+        if removeAction:
+            Utils.Print("removing {} from the security group".format(removeAction))
+            trans = self.defaultNode.pushMessage(self.contractAccount.name, "remove", removeAction, "--permission eosio@active")
+            Utils.Print("remove trans: {}".format(json.dumps(trans, indent=4, sort_keys=True)))
+
+        self.publishProcessNum += 1
```

Comment:
do we really need this functionality? We deal with small amount of participants and can simply publish 0 every time

Reply:
except then we would have identical transactions because there is nothing distinguishing one publish action from another. If we don't do this, we would need to add a nonce or else make sure we never send 2 publish actions in the same block

Reply:
that is known limitation to transactions. I think we just should never send two publish actions in one block. Was it ever issue to you or it was just in case?